### PR TITLE
Add feature Create DRG and attach to VCN

### PR DESCRIPTION
**Pull request by kral2**
Fix Issue #16 (Add a DRG and associate it with the VCN)
Fix Issue #19 (CHANGELOG.md is missing v2.0.0 information)
Fix Issue #15 (would be possible to avoid => REMOVE , ALL default values in any of the modules ?)

Add Deprecation notice for variable names (Issue #24)
Example gets the values from variables

suggested tag for release: v2.2.0

### CODE REVIEW

**Comment by kral2** [comment]
@hyder ready for review.

**Review by hyder — COMMENTED** [review]
reviewed

File: examples/main.tf [hunk 1]
```diff
@@ -4,31 +4,19 @@
...
   source = "../"
 
-  region = "us-phoenix-1"
-
   # general oci parameters
-
-  compartment_id = ""
-
-  label_prefix = "dev"
+  compartment_id = var.tenancy_id
```
[the remark this answers is not in the document]
Comment (kral2) [reply examples/main.tf:8]: The logic behind this was that `tenancy_id` is always set in the process to authenticate to the OCI API, while `compartment_id` is not necessarily set and we cannot determine a value ahead of time.

The only known value for compartment regardless of the environment is the tenancy ocid :-(

I agree creating any resource in the root compartment of a tenancy should be generally avoided, but that's ultimately the module user's choice and responsibility.

If the goal of the example is to have a resource deployed with the least amount of friction, I would prefer to have a copy/paste experience that works with minimum effort, and then tweak it to my needs (e.g. provide a specific compartment). Also `compartment_id` is updatable, so user can adjust the final result if needed.

Would it be a better solution if we write the example code in a way that works out of the box, and add a comment to advise setting something more relevant values for user's context?

```suggestion
  compartment_id = var.tenancy_id # please set `var.compartment_id` value and change `var.tenancy_id` to `var.compartment_id` to avoid creating resources in the root compartment
```

File: examples/variables.tf [hunk 2]
```diff
@@ -62,9 +64,25 @@ variable "service_gateway_enabled" {
   type        = bool
 }
 
+variable "create_drg" {
+  description = "whether to create Dynamic Routing Gateway. If set to true, creates a Dynamic Routing Gateway."
+  type        = bool
+  default     = true
```
[the remark this answers is not in the document]
Comment (kral2) [reply examples/variables.tf:70]:
```suggestion
  default     = false
```

Should I also set `lockdown_default_seclist` to `false` as default value in the example code?

File: examples/variables.tf [hunk 3]
```diff
@@ -62,9 +64,25 @@ variable "service_gateway_enabled" {
...
+  type        = string
+  default     = "hub"
+}
+
 variable "tags" {
   description = "simple key-value pairs to tag the resources created"
-  type = map(any)
+  type        = map(any)
+  default = {
+    Terraformed = "Yes"
```
[the remark this answers is not in the document]
Comment (kral2) [reply examples/variables.tf:83]:
```suggestion
    Terraformed = "yes"
```

File: variables.tf [hunk 4]
```diff
@@ -71,3 +75,16 @@ variable "vcn_name" {
...
+
+# drg parameters
+variable "create_drg" {
+  description = "whether to create Dynamic Routing Gateway. If set to true, creates a Dynamic Routing Gateway."
+  type = bool
+  default     = false
+}
+
+variable "drg_display_name" {
+  description = "(Updatable) Name of Dynamic Routing Gateway. Does not have to be unique."
```
[the remark this answers is not in the document]
Comment (kral2) [reply variables.tf:87]: That's partial a copy/paste from the resource documentation, but I think that's a valuable information for the user.

I am in favor to remove (Optional) or (Required) text, as it is detected by the registry and we can definitely generate documentation table with the correct status.

On the other hand, knowing if an input is updatable or not requires the user to check the original resource page. I consider having it here handy is a time saver.

File: CHANGELOG.adoc [hunk 5]
```diff
@@ -7,12 +7,39 @@ All notable changes to this project are documented in this file.
...
+* module example use variables
+
+=== Deprecation notice
+
+The folowwing variables will be renamed at the next major release of this module (related to Issue #24):
+
+* var.internet_gateway_enabled --> var.create_internet_gateway
+* var.nat_gateway_enabled --> var.create_nat_gateway
+* var.service_gateway_enabled --> var.create_service_gateway
+* var.tags --> var.create_freeform_tags
```
[the remark this answers is not in the document]
Comment (kral2) [reply CHANGELOG.adoc:23]:
```suggestion
* var.tags --> var.freeform_tags
```

File: variables.tf [hunk 6]
```diff
@@ -71,3 +75,16 @@ variable "vcn_name" {
   description = "user-friendly name of to use for the vcn to be appended to the label_prefix"
   type        = string
 }
+
+# drg parameters
+variable "create_drg" {
+  description = "whether to create Dynamic Routing Gateway. If set to true, creates a Dynamic Routing Gateway."
+  type = bool
```
[the remark this answers is not in the document]
Comment (kral2) [reply variables.tf:82]: will run a `terraform fmt` to fix these issues 👍

File: examples/variables.tf [hunk 7]
```diff
@@ -62,9 +64,25 @@ variable "service_gateway_enabled" {
...
+    Module      = "oracle-terraform-modules/vcn/oci"
```
[the remark this answers is not in the document]
Comment (kral2) [reply examples/variables.tf:84]:
```suggestion
    module      = "oracle-terraform-modules/vcn/oci"
```

File: examples/variables.tf [hunk 8]
```diff
@@ -35,11 +35,13 @@ variable "user_id" {
 variable "compartment_id" {
   description = "compartment id where to create all resources"
   type        = string
+  default     = null
```
[the remark this answers is not in the document]
Comment (kral2) [reply examples/variables.tf:38]: The value set here is relative to the behavior explained on the next review comment.

Having `null` value for `compartment_id` allows the user to execute the example code with minimal edit: just the oci provider authentication variables.

If the user copy/paste his existing `terraform.tfvars` file where `compartment_id` value is set, this will effectively overwrite this default value. Seems to me it is best of both worlds.

File: examples/variables.tf [hunk 9]
```diff
@@ -35,11 +35,13 @@ variable "user_id" {
...
 }
 
 variable "label_prefix" {
   description = "a string that will be prepended to all resources"
   type        = string
+  default     = "module-vcn"
```
[the remark this answers is not in the document]
Comment (kral2) [reply examples/variables.tf:44]: As discuted offline, we will prefer "terraform-oci" to the final name will be "terraform-oci-vcn" which indicates this vcn was created by the module, without being specific for a use case like "dev".

```suggestion
  default     = "terraform-oci"
```

File: examples/main.tf [hunk 10]
```diff
@@ -4,31 +4,19 @@
...
+  internet_gateway_enabled = var.internet_gateway_enabled
+  nat_gateway_enabled      = var.nat_gateway_enabled
+  service_gateway_enabled  = var.service_gateway_enabled
+  create_drg               = var.create_drg
+  drg_display_name         = var.drg_display_name
+  tags                     = var.tags
+  vcn_cidr                 = var.vcn_cidr
+  vcn_dns_label            = var.vcn_dns_label
+  vcn_name                 = var.vcn_name
+  lockdown_default_seclist = var.lockdown_default_seclist
```
[the remark this answers is not in the document]
Comment (kral2) [reply examples/main.tf:21]: Reordering done in both`main.tf` and `variables.tf` for /examples:
- variables.tf list variables in alphanumerical ordering for each section
- module block in `main.tf` list inputs in the same group/definition order as in `variables.tf`.

File: examples/variables.tf (hunk 9 above)
[the remark this answers is not in the document]
Comment (kral2) [reply examples/variables.tf:44]: I also set default values for `var.vcn_name` and `var.vcn_dns_label` in `examples/variables.tf`

Final result when applying the configuration template from `/examples` without any changes will be:
- vcn display name : terraform-oci-vcn-modules
- vcn dns domain name : vcnmodule.oraclevcn.com

File: variables.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (hyder) [reply variables.tf:89]: Ok

File: examples/main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (hyder) [reply examples/main.tf]: I don't agree with this. We use the tenancy_id mainly for initializing providers whereas compartment_id to create resources. 

Yes, technically root compartment is the tenancy id but and I know it is possible to create stuff in root compartment but that's promoting bad practices and we shouldn't be doing that.  Users should explicitly enter compartment_id of where they want their resources created. If they don't root = tenancy ids but the need to pass a compartment id forces them to create one anyway, then we are guiding them towards good practices.

File: examples/variables.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (hyder) [reply examples/variables.tf:38]: No, please don't this as null. Please request the user to explicitly pass their compartment id.

File: examples/main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (kral2) [reply examples/main.tf]: OK. I guess this decision will make his way into the codingconventions.adoc 👍

File: examples/variables.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (kral2) [reply examples/variables.tf:38]: OK, I guess we can add information on the example usage listing which variables **must** be set by the user before using the example.

That's basically anything related to auth + compartment.

File: vcn_gateways.tf [hunk 11]
```diff
@@ -89,3 +98,23 @@ resource "oci_core_route_table" "nat" {
 
   count = var.nat_gateway_enabled == true ? 1 : 0
 }
+
+###############################
+# Dynamic Routing Gateway (DRG)
+###############################
+
+resource "oci_core_drg" "this" {
```
[the remark this answers is not in the document]
Comment (kral2) [reply vcn_gateways.tf:106]: or only "drg" similar to internet gateway that is named "ig"?
drg is already an acronym containing "gateway" in it.

**Review by hyder — APPROVED** [review]
Approved

**Comment by Djelibeybi** [comment]
BTW, it's generally considered more polite to have a pull request per issue so that reviewers can focus on whether the proposed changes fix that one problem. With multiple fixes in a single PR, reviewers have to context shift each change to the issue it's addresing which makes things a lot more complex.

**Comment by kral2** [comment]
> BTW, it's generally considered more polite to have a pull request per issue so that reviewers can focus on whether the proposed changes fix that one problem. With multiple fixes in a single PR, reviewers have to context shift each change to the issue it's addresing which makes things a lot more complex.

Agree. Some actions triggered during PR review should have been handled as new issues and separate PR. This do not reflect anymore in the PR as I have squashed my commits.

I will try to keep PR subject more focused to facilitate code review 👍

File: docs/prerequisites.adoc [hunk 12]
```diff
@@ -1,47 +1,35 @@
...
-. link:#using-autonomous-linux[Using Autonomous Linux]
+toc::[]
 
-=== Install Terraform
+== Install Terraform
 
 Start by installing Terraform and configuring your path.
 
-==== Installing Terraform on Oracle Linux
+=== Installing Terraform on Oracle Linux
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:32]: This would only work on Oracle Linux 7. 

For Oracle Linux 8, you'd need to run `dnf -y install oraclelinux-developer-release-el8 && dnf -y install terraform`

File: docs/prerequisites.adoc [hunk 13]
```diff
@@ -50,16 +38,18 @@ If you are using Oracle Linux, you can install Terraform as follows:
...
 If you are using a Mac, you can install Terraform as follows:
 
 [source,bash]
 ----
-brew install terraform
+brew install hashicorp/tap/terraform
 ----
 
-==== Manual Installation
+This will install Terraform using the official Hashicorp homebrew tap. For more details between the differences and benefits to use the community maintained or the official homebrew tap, see {uri-hashicorp-brew-tap-announcement}[Hashicorp's announcement]
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:50]: Homebrew is a product name and should be capitalised. HashiCorp should be camed-cased.

File: docs/prerequisites.adoc [hunk 14]
```diff
@@ -50,16 +38,18 @@ If you are using Oracle Linux, you can install Terraform as follows:
 yum -y install oraclelinux-developer-release-el7 && yum -y install terraform
 ----
 
-==== Installing Terraform on macOS
+=== Installing Terraform on macOS
 
 If you are using a Mac, you can install Terraform as follows:
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:43]: I would reword this to "If you are using macOS, you can use Homebrew to install Terraform"
 with a link to https://brew.sh/. Keep in mind that Homebrew isn't installed by default on macOS, so you need to include that as a step.

File: docs/prerequisites.adoc [hunk 15]
```diff
@@ -50,16 +38,18 @@ If you are using Oracle Linux, you can install Terraform as follows:
...
-brew install terraform
+brew install hashicorp/tap/terraform
 ----
 
-==== Manual Installation
+This will install Terraform using the official Hashicorp homebrew tap. For more details between the differences and benefits to use the community maintained or the official homebrew tap, see {uri-hashicorp-brew-tap-announcement}[Hashicorp's announcement]
+
+=== Manual Installation
 
 If you prefer to install Terraform manually, you can follow these instructions:
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:54]: You don't need to provide manual instructions here. Just link to the https://www.terraform.io/downloads.html page, i.e. "For manual installation, follow the instructiuons on the Terraform download page."

File: docs/prerequisites.adoc [hunk 16]
```diff
@@ -80,7 +70,7 @@ Open a terminal and type the following:
 sudo mv /path/to/terraform /usr/local/bin
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:70]: Never move binaries. That's not how you configure the path. Either update `.bashrc` with something like `export PATH=/path/to/terraform/bin:$PATH` or create a symbolic link  (which is what Homebrew does).

File: docs/prerequisites.adoc [hunk 17]
```diff
@@ -90,27 +80,29 @@ Follow the steps below to configure your path on Windows:
 . Close all open windows by clicking 'OK'
 . Open a new terminal and verify terraform has been properly installed
 
-==== Testing Terraform installation
+=== Testing Terraform installation
 
-Open a terminal and test:
+Open a terminal type this command:
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:85]: You're missing the word "and"

File: docs/prerequisites.adoc [hunk 18]
```diff
@@ -90,27 +80,29 @@ Follow the steps below to configure your path on Windows:
...
 
 [source,bash]
 ----
-terraform -v
-Terraform v0.12.24
+> terraform -v
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:89]: Always provide the expected/sample output, i.e.

```shell
$ terraform -v
Terraform v0.14.7
```

File: docs/prerequisites.adoc [hunk 19]
```diff
@@ -90,27 +80,29 @@ Follow the steps below to configure your path on Windows:
...
 ----
 
-=== Generate and upload your OCI API keys
+The first line of the output should be `Terraform vX.X.X` with your installed version.
+You will also see below any provider that is accessible to Terraform in the current folder.
+
+== Generate and upload your OCI API keys
 
 Follow the documentation for {uri-oci-keys}[generating and uploading your api keys].
```
Comment (Djelibeybi) [inline docs/prerequisites.adoc:97]: API should be all capitals as it is an acronym.

File: docs/quickstart.adoc [hunk 20]
```diff
@@ -16,23 +15,21 @@
...
 
 1. You have setup the {uri-oci-keys}[required keys]
 2. You know the {uri-oci-ocids}[required OCIDs]
 3. You have the necessary permissions
 
-=== Pre-requisites
+== Pre-requisites
 
 1. git is installed
 2. ssh client is installed
```
Comment (Djelibeybi) [inline docs/quickstart.adoc:28]: SSH should be all capitals as it is an acronum.

File: docs/quickstart.adoc [hunk 21]
```diff
@@ -16,23 +15,21 @@
...
+== Assumptions
 
 1. You have setup the {uri-oci-keys}[required keys]
 2. You know the {uri-oci-ocids}[required OCIDs]
 3. You have the necessary permissions
 
-=== Pre-requisites
+== Pre-requisites
 
 1. git is installed
```
Comment (Djelibeybi) [inline docs/quickstart.adoc:27]: Git should be capitalised

File: examples/README.md [hunk 22]
```diff
@@ -88,13 +90,13 @@ module "vcn" {
 
 1. Create your own module e.g. subnets. In modules directory, create a subnets directory:
 
-```
+```Shell
```
Comment (Djelibeybi) [inline examples/README.md:93]: This should be ```shell

File: examples/main.tf [hunk 23]
```diff
@@ -1,34 +1,47 @@
 # Copyright (c) 2019, 2021, Oracle Corporation and/or affiliates.
 # Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
 
+# Versions
```
Comment (Djelibeybi) [inline examples/main.tf:4]: I would say "Specify version requirements"

**Review by Djelibeybi — CHANGES_REQUESTED** [review]

File: docs/prerequisites.adoc (hunk 16 above)
Comment (kral2) [reply docs/prerequisites.adoc:70]: I think the `configure your path` sections should be simply removed. If the user decides to go with manual installation, they can follow the instruction from terraform page as you suggested and manage their operating system as they prefer.

Terraform install page simply says the following, with a link to StackExchange for information related to `PATH`.

>Terraform is distributed as a single binary. Install Terraform by unzipping it and moving it to a directory included in your system's PATH .

File: docs/prerequisites.adoc (hunk 18 above)
Comment (kral2) [reply docs/prerequisites.adoc:89]: I think I will remove this section:
- that's part of the installation process
- output being a version, it will quickly become obsolete and ask for further maintenance :-)

File: CHANGELOG.adoc [hunk 24]
```diff
@@ -3,21 +3,71 @@
...
+** added outputs for cli feedback
+* added oci provider minimum version to 4.0.0
+* documentation update: coding conventions
+
+=== Fixes
+* added CHANGELOG entry for v2.0.0 (fixes #19)
+
+=== Deprecation notice
+
+The folowwing variables will be renamed at the next major release of this module (related to Issue #24):
```
Comment (Djelibeybi) [inline CHANGELOG.adoc:37]: Typo: following. And "issue" should be lowercase.

File: CONTRIBUTING.adoc [hunk 25]
```diff
@@ -1,14 +1,14 @@
...
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+endif::[]
+:toc:
 
 :uri-oracle-oca: https://www.oracle.com/technetwork/community/oca-486395.html
```
Comment (Djelibeybi) [inline CONTRIBUTING.adoc:11]: Please change this to our new online OCA which is now live: https://oca.opensource.oracle.com

File: CONTRIBUTING.adoc [hunk 26]
```diff
@@ -31,18 +31,18 @@ Please ensure the documentation is updated to include any new feature, functiona
 
 You should ensure that your documentation changes include the following:
 
-- Update {uri-changelog}[CHANGELOG] under the 'Unreleased' section. The maintainers will include this to the appropriate version section during the release process for that version
+- Update xref:CHANGELOG.adoc[CHANGELOG] under the 'Unreleased' section. The maintainers will include this to the appropriate version section during the release process for that version
```
Comment (Djelibeybi) [inline CONTRIBUTING.adoc:34]: Perhaps: "Add an entry that explains your fix or enhancement to the list of unreleased items. The maintainers will move your entry to the appropriate version during the release process for that version."

File: README.adoc [hunk 27]
```diff
@@ -40,6 +40,9 @@ It creates the following resources:
 * An optional internet gateway
 * An optional NAT gateway
 * An optional service gateway
+* An optional Dynamic Routing Gateway
```
Comment (Djelibeybi) [inline README.adoc:43]: The other items in the list are lowercase, so dynamic routing gateway should be too.

File: README.md [hunk 28]
```diff
@@ -8,6 +8,7 @@ It creates the following resources:
 * An optional internet gateway
 * An optional NAT gateway
 * An optional service gateway
+* An optional Dynamic Routing Gateway
```
Comment (Djelibeybi) [inline README.md:11]: Same as above: dynamic routing gateway should be lowercase.

File: docs/codingconventions.adoc [hunk 29]
```diff
@@ -21,8 +31,75 @@ It is not an exhaustive list and it will be updated as we go. In the mean time,
...
+== Documentation format
+
+We use https://asciidoc.org/[AsciiDoc] with the `*.adoc` file extension. The only exception is for the README files that need to be displayed on the Terraform Registry.
+
+[NOTE]
+====
+As GitHub renders AsciiDoc files using https://asciidoctor.org/[AsciiDoctor], it is the current reference regarding AsciiDoc syntax we use in the documentation.
+====
+
+Terraform Registry currently do not support AsciiDoc,so documentation content such as `README` files must be in Markdown format.
```
Comment (Djelibeybi) [inline docs/codingconventions.adoc:43]: Try: "As the Terraform Registry does not support AsciiDoc, the `README.md` and certain other documentation must be available in Markdown format,"

File: docs/codingconventions.adoc [hunk 30]
```diff
@@ -21,8 +31,75 @@ It is not an exhaustive list and it will be updated as we go. In the mean time,
...
+link:examples/README.md[examples README]
+```
+
+=== External links
+
+For external links, we use https://docs.asciidoctor.org/asciidoc/latest/macros/autolinks/[autolinks].
+
+[TIP]
+====
+If a link is used mutiple times, prefer to add the link in the header with a reference in the format `:uri-xxx:`. This will facilitate updates in link destination is to be modified.
```
Comment (Djelibeybi) [inline docs/codingconventions.adoc:65]: Try: "Links that are used multiple times must be defined in the header using the `:uri-xxx` reference format. This makes any future updates to the URI much simpler."

File: docs/codingconventions.adoc [hunk 31]
```diff
@@ -21,8 +31,75 @@ It is not an exhaustive list and it will be updated as we go. In the mean time,
...
+
+----
+:uri-repo: https://github.com/oracle-terraform-modules/terraform-oci-vcn
+
+{uri-repo}[link caption]
+----
+
+====
+
+===  Terraform registry specificities
```
Comment (Djelibeybi) [inline docs/codingconventions.adoc:78]: Terraform Registry requirements

**Review by Djelibeybi — CHANGES_REQUESTED** [review]

File: README.md (hunk 28 above)
Comment (kral2) [reply README.md:11]: changed in both adoc and md version of the file

**Review by Djelibeybi — APPROVED** [review]
Let's merge this PR now and we can tweak any minor issues in future PRs. :)
